Plexil interface for ArmMoveJoint and ArmMoveJoints actions

MilanAdd opened the pull request:
The new ROS actions ArmMoveJoint and ArmMoveJoints can now be commanded from PLEXIL.

To test:

- Launch any simulator of choice
- Then, on a separate terminal, execute the test plans, whatever order you may choose, for ArmMoveJoint and ArmMoveJoints (respectively) by the commands `roslaunch ow_plexil ow_exec.launch plan:="TestArmMoveJoint.plx"` and `roslaunch ow_plexil ow_exec.launch plan:="TestArmMoveJoints.plx"`

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoint.plp:
```diff
@@ -0,0 +1,28 @@
+  LibraryCall ArmMoveJoint (Relative=true, Joint=6, Angle=0.0);
+
+  log_info ("ArmMoveJoint demo finished.");
+}
```
This sequential test looks good, after you change any Angle settings that aren't working -- I've asked @anjan1984 to look into valid ranges for each joint.

There are two more tests you could add (in this plan) to better test this action: a sequence of successive angles commanded to one (any) joint, and a Concurrence of all 6 joints commanded.

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoints.plp:
```diff
@@ -0,0 +1,29 @@
+TestArmMoveJoints:
+{
+  // Three sets of angles for each joint
+  Real angles1[6] = #(0.0 -0.5 0.0 2.0 0.0 0.0);
+  Real angles2[6] = #(0.0 0.5 0.0 0.0 0.0 0.0);
+  Real angles3[6] = #(0.0 -0.5 0.0 0.0 0.0 0.0);
```
For more interesting tests, I suggest non-zero angles for all Relative moves.

kmdalal commented on ow_plexil/src/plexil-adapter/OwInterface.h:
```diff
@@ -123,6 +134,10 @@ class OwInterface : public PlexilInterface
   void guardedMoveAction (double x, double y, double z,
                           double dir_x, double dir_y, double dir_z,
                           double search_distance, int id);
+  void armMoveJointAction (bool relative, int joint, 
+                     double angle, int id);
+  void armMoveJointsAction (bool relative, const std::vector<double>& angles,
+                      int id);
```
Minor: the second line of each declaration needs some indentation to line up.

kmdalal commented:
@MilanAdd , the code looks good -- just some small comments and suggestions for improvement.   

Unfortunately I have not yet been able to test this because my Ubuntu machine has an overheating problem and crashes whenever I start Gazebo.  I hope to have this resolved very soon, and I will then try your plans as they are, and after you correct the errors.

kmdalal commented on ow_plexil/src/plans/ArmMoveJoint.plp:
```diff
@@ -0,0 +1,37 @@
+ArmMoveJoint:
+{
+  In Boolean Relative;
+  In Integer Joint;
```
Would be good to check that Joint is in the range 0-5, and issue an error otherwise.  Once we learn the range of each joint, those could be checked here as well.

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoint.plp:
```diff
@@ -0,0 +1,28 @@
+
+  LibraryCall ArmMoveJoint (Relative=true, Joint=5, Angle=0.0);
+
+  LibraryCall ArmMoveJoint (Relative=true, Joint=6, Angle=0.0);
```
Valid range for joint index is 0..5.  This line causes an error.

kmdalal replied:
Valid range for joint index is 0..5.  This line causes an error.

kmdalal commented:
I was able to test, and got the same errors you sent me.   One of the errors is from commanding joint index 6 in a test plan -- see the comment there.  The rest are from attempting invalid angles.  I experimented and found that 0.0 is valid for all joints, but 0.1 is not valid for one or two joints.  Hopefully @anjan1984 can provide some pointers on knowing valid joint ranges.

Another important improvement to the test plans (after finding joint angles that work) is making them re-runnable.  To do that, you should move a joint to an angle, and then back to the angle where it started (which could be 0.0).   Otherwise, a second run of the same plan won't have any effect.

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoint.plp:
```diff
@@ -0,0 +1,90 @@
+  
+  
+  // Concurrence is added to all six joints commanded
+  TestArmMoveJointConc: Concurrence
```
I presume the Concurrence is not working for you.  I discovered something I had not anticipated.  The framework for actions allows only one instance of any lander operation to be run at a time.  This is the first situation where multiple instances of the same operation (but with different joints) makes sense.   

I think it's best to omit this test.  We could create a Jira to support multiple instances of ArmMoveJoint, but it's better and easier to just use ArmMoveJoints.

MilanAdd replied:
Yeah, concurrence was not working for me either and you do have a good point about just using ArmMoveJoints for any concurrent tests. I will remove it then.

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoints.plp:
```diff
@@ -0,0 +1,26 @@
+  log_info ("Starting TestArmMoveJoints plan...");
+
+  LibraryCall ArmMoveJoints (Relative=false, Angles=angles1);
+
```
This call produced an error in the simulator:

`ValueError: operands could not be broadcast together with shapes (6,) (0,) `

This needs some debugging, but I suspect that the vectors you define here are not getting correctly converted into the vectors that the Python-based action client gets (this error comes from Python).  I can execute these three commands manually using the Python client scripts without any problems.

Try inspecting (e.g. printing out) the array that gets down to `OwInterface::armMoveJoints` and see if you can find the source of the problem.   If the array looks good there, have a look at what's coming into `ow_simulator/ow_lander/scripts/arm_action_servers.py`

kmdalal commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -869,6 +885,66 @@ void OwInterface::guardedMoveAction (double x, double y, double z,
+
+  ArmMoveJointsGoal goal;
+  goal.relative = relative;
+  for(int i = 0; i < goal.angles.size(); i++){
+    goal.angles[i] = angles[i];
+  }
```
Continuing our email discussion, I'm guessing the problem somehow starts here, and that the array getting assigned to `goal.angles` (or not) is not the one coming in (`angles`) which looks correct.  I tried a few easy experiments, and did a quick web search on this topic, no luck.

kmdalal replied:
I did some experimentation and am committing a simple solution.  One reason the existing code didn't work is that goal.angles.size() is unspecified at first -- the ROS array has no defined size.  I tried doing a resize(6) but this didn't work for some reason.

kmdalal commented on ow_plexil/src/plans/ArmMoveJoints.plp:
```diff
@@ -0,0 +1,48 @@
+    SynchronousCommand arm_move_joints (Relative, Angles);
+  }
+
+  SendArmMoveJointsParallel: Concurrence
+  {
+    Start !Lookup(ArmFault);
+
+    if FaultDetected
+    {
+      log_info ("Arm fault(s) resolved, sending arm_move_joints command to lander...");
+    }
+    
+    SynchronousCommand arm_move_joints (Relative, Angles);
+  }
```
Perhaps this node is here accidentally, but it doubles the operation and causes problems.

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoint.plp:
```diff
@@ -0,0 +1,70 @@
+
+  
+  // Moves a particular joint to a sequence of successive angles, 
+  // making this sequence of plan re-runnable by moving joint 
+  // back to angle 0.0 before it moves to another angle
+  // each time
```
I added a Stow() before and after the first test to achieve re-run-ability in cleaner way.

kmdalal commented on ow_plexil/src/plans/TestArmMoveJoint.plp:
```diff
@@ -0,0 +1,70 @@
+  {
+    log_info ("Starting first ArmMoveJoint test...");
+  
+    LibraryCall ArmMoveJoint (Relative=true, Joint=0, Angle=-1.0);
```
This angle doesn't work (I checked it manually).  Changed to 0.

kmdalal commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -869,6 +885,66 @@ void OwInterface::guardedMoveAction (double x, double y, double z,
+{
+  ArmMoveJointGoal goal;
+  goal.relative = relative;
+  goal.joint = joint;
```
Subtle problem here.  goal.joint is really an int64 (long int).  The assignment is safe since no precision is lost, but I added a comment.

kmdalal commented:
I've added a bunch of comments on the files, AND committed the fixes.  Most were either subtle issues that took some experimentation, or simple ones that were trivial to fix.   @MilanAdd let me know if you have questions about any of these changes.  Try both the test plans and see if they run without any errors.

kmdalal commented:
It looks like my commit wiped out a number of the comments I left (e.g. because they were on lines that were removed).  Let's go over my changes in person on Monday.

MilanAdd commented:
> I've added a bunch of comments on the files, AND committed the fixes. Most were either subtle issues that took some experimentation, or simple ones that were trivial to fix. @MilanAdd let me know if you have questions about any of these changes. Try both the test plans and see if they run without any errors.

I tested out both test plans and they now successfully work without any major errors, especially ArmMoveJoints, which has worked flawlessly. For ArmMoveJoint, I always get these error messages on the simulator terminal (those have always been there):
```
[ERROR] [1659372167.628405885, 1916524938.067500114]: arm/arm: Unable to sample any valid states for goal tree
[ WARN] [1659372167.628769043, 1916524938.067500114]: Fail: ABORTED: No motion plan found. No execution attempted.
```

It has very little impact on the execution of the test plan, but I'm still curious on how to fix it, but again this is pretty trivial. Other than that, the changes you made have been pretty successful in terms of fixing some of the issues that have occurred when executing these test plans, so I truly appreciate your help and guidance. I just added some more Stow commands to make the test plans even more re-runnable.

kmdalal commented:
Great, we're getting there!  That last remaining error is one to be tracked down.  See if you can pinpoint which call toe ArmMoveJoint is causing it (I added INFO statements that show each call), and then reproduce the situation manually using the ow_simulator/ow_lander/scripts/arm_move_joint_client.py.  If you can reproduce it that way (which I hope), experiment with different values until you find one that works.   @anjan1984 is still working on establishing the joint ranges.

In any case, let's meet online sometime soon so that I can explain the changes I made.  I would not have expected you to find/solve some of them - they were a bit tricky even for me.

kmdalal reviewed (approved):
Great work!
